sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -322,6 +323,62 @@ class Analyzer(
   }
 
   /**
+    * Replace the index with the related attribute for order by and group by.
+    */
+  object ResolveIndexReferences extends Rule[LogicalPlan] {
+
+    def indexToColumn(
+        sortOrder: SortOrder,
+        child: LogicalPlan,
+        index: Int,
+        direction: SortDirection): SortOrder = {
+      val orderNodes = child.output
+      if (index > 0 && index <= orderNodes.size) {
+        SortOrder(orderNodes(index - 1), direction)
+      } else {
+        throw new UnresolvedException(sortOrder,
+          s"""Order by position: $index does not exist \n
+              |The Select List is indexed from 1 to ${orderNodes.size}""".stripMargin)
+      }
+    }
+
+    def indexToColumn(
```

Comment:
it indeed similar from the code structure and name, but internally it wants to construct two different things. And also the first parameter of `UnresolvedException` require concrete type, so it cannot be a type like TreeNode[_] to unify the api. How about I rename those to be `indexToSortOrder` and `indexToAggregate` ?